feat(usescrollsync): add scroll syncing via hook: `useScrollSync`

pcattori opened the pull request:
@okonet : I took a shot at https://github.com/okonet/react-scroll-sync/issues/36#issuecomment-639553613 .

Right now I wrote this as a stand-alone, typed React hook. Ideally, if we agree on the hook, we could expose it as a feature via `import {useScrollSync} from 'react-scroll-sync'`, but additionally the existing `ScrollSync` and `ScrollSyncPane` component could optionally be refactored to make use of `useScrollSync` (perhaps via context API).

**Looking for feedback** to see if this is on the right path before continuing to polish this PR.

[CodeSandbox Example](https://codesandbox.io/s/usescrollsync-9j2ox?file=/src/useScrollSync.ts) : syncing scroll of the red header and the white body on horizontal axis.

Relates to #22, #36

---

Example usage:

```tsx
const MyComponent: React.FC = () => {
  const [thead, theadRef] = useElement()
  const [tbody, tbodyRef] = useElement()
  useScrollSync([thead, tbody], { axis: 'horizontal' })
  return (
    <>
      <div ref={theadRef} role="rowgroup" style={{ overflowX: 'auto' }}>{...}</div>
      <div ref={tbodyRef} role="rowgroup" style={{ overflow: 'auto' }}>{...}</div>
    </>
  )
}
```

Other changes:
- `axis` (typed as `'horizontal' | 'vertical' | 'both'`) to replace `horizontal=true` and `vertical=true` options. This allows user to only enable the axes they _want_, as opposed to having to disable the axes they don't want.
- Removed concept of "groups"; instead just use multiple calls of `useScrollSync`, one for each group.
- Included a `useElement` hook for convenience. It just copies the [recommended approach from the React docs](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node) and wraps it in a hook.

okonet commented:
I like this direction very much! Do you want to drop components in favor of hooks and do a breaking change? Or do you want to support both?

okonet commented on src/example-useScrollSync.tsx:
```diff
@@ -0,0 +1,146 @@
+
+import { useScrollSync } from "react-scroll-sync";
+
+const App: React.FC = () => {
```
No need to type it. It’s discouraged by React team afaik.

pcattori replied:
Good point! Especially since `App` doesn't make use of `children`, which `React.FC` implies.

So I agree! 😄 

Out of curiosity, do you have a link to the React team discouraging this? I'm interested in how they phrased it.

okonet replied:
I can't remember where I read about it TBH.

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,91 @@
+import { useEffect } from "react";
+
+interface Options {
+  proportional?: boolean;
```
Maybe also an enum? “proportional” | “absolute”?

pcattori replied:
Something like `mode: 'proportional' | 'absolute'`?

Does it make sense to default this to `'proportional'` or `'absolute'`? I know the current implementation defaults to `proportional=true`, but I feel like `'absolute'` might be a more natural default.

Actually, I might be most in favor of not defaulting this to anything at all and requiring user to specify.

okonet replied:
Yes, exactly. I think defaulting to `proportional` is more appropriate but I'm open for making it a required prop.

pcattori replied:
Made `mode` a required parameter for now.

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,91 @@
+  useEffect(() => {
+    const nodes = refs
+      .map((ref) => ref.current)
+      .filter((n) => n !== null) as HTMLElement[];
```
One loop should be enough here. No need for this map. Just access the `.current` directly.

pcattori replied:
I actually have a local implementation that makes the ref handling much simpler and should address this. Namely, the `useScrollSync` hook will just accept a list of `HTMLElement` (rather than a list of refs to `HTMLElement`s).

I'll update with this change soon.

pcattori replied:
I've updated the PR to include this new implementation. For now just adding as another commit, but if we get to the point where we want to merge it, I'm happy to squash/rebase commits as needed.

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,91 @@
+    const nodes = refs
+      .map((ref) => ref.current)
+      .filter((n) => n !== null) as HTMLElement[];
+    nodes.map((node) => {
```
Should be a `forEach` since you not returning.

pcattori replied:
done

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,91 @@
+        node.onscroll = null;
+      });
+    };
+  }, []);
```
Missing dependencies here. When refs will change this effect won’t run.

pcattori replied:
done

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,91 @@
+  followers: HTMLElement[],
+  options: Options
+) => {
+  followers.map((follower) => {
```
Same as before: `forEach`

pcattori replied:
done

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,91 @@
+  if ((axis === "both" || axis === "horizontal") && scrollLeftOffset > 0) {
+    follower.scrollLeft = proportional
+      ? (paneWidth * scrollLeft) / scrollLeftOffset
+      : scrollLeft; // eslint-disable-line
```
Why is linter has to be disabled?

pcattori replied:
Not sure actually. I copied most of this from the [existing implementation](https://github.com/okonet/react-scroll-sync/blob/master/src/ScrollSync.js#L119). I tried removing the `eslint-disable-line`, but my editor keeps adding it back in... I'll investigate to see if it's just my editor config or something else.

pcattori replied:
🤷 not sure why the linter was being disabled before. I removed the `eslint-disable-line` comment and things seem to do just fine. so this should now be addressed.

okonet reviewed (changes requested):
Good direction!

pcattori commented:
> I like this direction very much! Do you want to drop components in favor of hooks and do a breaking change? Or do you want to support both?

I think it makes sense to keep the components, at least temporarily. Mostly because there may be use-cases that rely on features from the component implementation (e.g. I don't implement the `onSync` callback yet).

Options:
1. Keep the components. Support both components and hooks.
2. Same as (1), but deprecate the components. This will give us some time to get feedback for bugs/features for the hook implementation.
3. Remove the components.

I think I lean towards (2). What do you think @okonet ?

okonet reviewed (commented):
Sorry I lost the track on this one. It looks good now so let's resolve the last comment. Also, the documentation needs to be updated.

okonet commented:
@pcattori I resolved almost all comments and awaiting for the final changes. Also, could you document the hook / update the documentation to reflect the new state?

I would say, let's remove components since maintenance efforts are way too high. Also documenting both approaches going to take longer. React isn't going back to pre-hook era anyways and we'll release this as a breaking change (we could even change the name of the project and package to reflect the change). WYT?

pcattori commented:
> @pcattori I resolved almost all comments and awaiting for the final changes. Also, could you document the hook / update the documentation to reflect the new state?
> 
> I would say, let's remove components since maintenance efforts are way too high. Also documenting both approaches going to take longer. React isn't going back to pre-hook era anyways and we'll release this as a breaking change (we could even change the name of the project and package to reflect the change). WYT?

@okonet sounds good! I'll take a look at this over the weekend to finalize things.

pcattori commented:
@okonet: I've implemented the `mode` parameter, added jsdoc for `useScrollSync`, and reworked the example into a markdown usage guide.

I think we should be good to go with that! Let me know if we need anything else before merging.

Next steps:
- Remove the existing components to decrease maintenance overhead

okonet commented on src/useScrollSync.ts:
```diff
@@ -0,0 +1,101 @@
+import React from "react";
+
+interface Options {
+  mode: "proportional" | "absolute";
```
I think it’s odd to make `Mode` a required option but `axis` not. I would suggest to make both options optional (hence the name) and use same defaults as the components do. Alternatively, if you think it’s better, let’s make both options required. WYT?

pcattori replied:
Yes, I agree. I've made the relevant changes now. Let me know what you think.

okonet reviewed (changes requested):
Let’s decide on both options to be options or required and then it’s good to go! Great work!

okonet commented:
Looking great! I'd say, let's release it as a minor version.

okonet commented:
Oh one last thing, we should add it to the documentation website.

For that we'd need to modify the glob in the https://github.com/okonet/react-scroll-sync/blob/master/styleguide.config.js to include the hook as well. WYT?

You can preview it here https://deploy-preview-54--react-sync-scroll.netlify.app

jayhuang-7 commented:
Really appreciate your work, everyone.
but it seems like some type checking in TS are not passing. `useElement` is returning `CallbackRef<T | undefined>` but it is not compatible with `RegacyRef<T | null>`.

pcattori commented:
> @okonet, @pcattori Really appreciate your work, everyone.
> but it seems like some type checking in TS are not passing. `useElement` is returning `CallbackRef<T | undefined>` but it is not compatible with `RegacyRef<T | null>`.

@jayhuang-7 : could you provide an example where this happens?

jayhuang-7 commented:
@pcattori 
I just followed the Readme.md like this one.

```javascript
const MyComponent: React.FC = () => {
  const [thead, theadRef] = useElement()
  const [tbody, tbodyRef] = useElement()
  useScrollSync([thead, tbody], { axis: 'horizontal' })
  return (
    <>
      <div ref={theadRef} role="rowgroup" style={{ overflowX: 'auto' }}>{...}</div>
      <div ref={tbodyRef} role="rowgroup" style={{ overflow: 'auto' }}>{...}</div>
    </>
  )
}
```
However, it seems that we can't set `theadRef` and `tbodyRef` in line 7 and 8 due to type error that I indicated the above.
